src/subscriptionNotifier/subscriptionNotifier.service.ts

```diff
@@ -0,0 +1,237 @@
+import { CabinService } from 'src/cabin/cabin.service';
+import { EmbedBuilder } from 'discord.js';
+
+type MessageThread = {
+  topic: string | null;
+  mainMessage: string;
+  embeds: EmbedBuilder[];
+};
+
+@Injectable()
+export class SubscriptionNotifierService {
+  private readonly logger = new Logger(SubscriptionNotifierService.name);
+  private task: cron.ScheduledTask;
+
+  constructor(
+    private readonly activityService: ActivityService,
+    private readonly cabinService: CabinService,
+    private readonly configService: ConfigService,
+    private readonly discordService: DiscordService,
+    private readonly subscriptionService: SubscriptionService,
+  ) {
+    if (this.configService.get('NODE_ENV') === 'production') {
+      this.task = cron.schedule('0 16 * * *', this.on4pm);
+      this.logger.log('setup 4pm subscription notifier');
+    }
+  }
+
+  on4pm = () => {
+    this.logger.log('on4pm');
+    this.notifySubscribers()
+      .catch((error) => {
+        this.logger.error(
+          'Sending subscription notifications failed with error.',
+          error,
+        );
+      })
+      .then(() =>
+        this.logger.log('Subscription notifications sent successfully.'),
+      );
+  };
+
+  private async notifySubscribers() {
+    const subscribers = await this.subscriptionService.getSubscribers();
+
+    subscribers?.forEach(async (subscriber) => {
+      const subscriptions = await this.subscriptionService.getSubscriptions(
+        subscriber.subscriberId,
+      );
+      if (subscriptions.length === 0) {
+        return;
+      }
+
+      const checkedAt = new Date();
+      const subscriptionsWithNews = await this.findNewsForSubscriptions(
+        subscriptions,
+      );
+      if (subscriptionsWithNews.length === 0) {
+        return;
+      }
+
+      const messages = this.createNotificationMessages(subscriptionsWithNews);
+      if (subscriber.subscriberType === 'channel') {
+        await this.sendNotificationsInThreads(subscriber, messages);
+      }
+      if (subscriber.subscriberType === 'user') {
+        await this.sendNotificationsAsDm(subscriber, messages);
+      }
+
+      await this.subscriptionService.updateNotifiedAt(subscriptions, checkedAt);
+    });
+  }
+
+  private async findNewsForSubscriptions(subscriptions: subscriptions[]) {
+    const subscriptionsWithNews = [];
+
+    for (const subscription of subscriptions) {
+      if (subscription.type === 'cabins') {
+        const news = await this.getNewCabins(subscription);
+        if (news.length > 0) {
+          subscriptionsWithNews.push({
+            subscription: subscription,
+            news: news,
+          });
+        }
+      }
+
+      if (subscription.type === 'activities') {
+        const news = await this.getNewActivities(subscription);
+        if (news.length > 0) {
+          subscriptionsWithNews.push({
+            subscription: subscription,
+            news: news,
+          });
+        }
+      }
```

Comment:
I am wondering if it would be cleaner to separate these out at the query level so the logic becomes:

1. Get all cabin subscriptions
2. Send cabin notifications
3. Get all activity subscriptions
4. Send activity notifications.


This means we don't need to have any conditionals related to the type.

Reply:
I'm not sure how this would get rid of the conditionals? You still have to treat them differently since the objects come from different tables, so notification texts, embeds etc. needs to be build differently?

Reply:
As they would be sent separately then you can have separate functions for each this should help avoid the conditionals.